src/java.base/share/classes/java/io/Reader.java

```diff
@@ -184,12 +184,34 @@ protected Reader(Object lock) {
      * @since 1.5
      */
     public int read(CharBuffer target) throws IOException {
-        int len = target.remaining();
-        char[] cbuf = new char[len];
-        int n = read(cbuf, 0, len);
-        if (n > 0)
-            target.put(cbuf, 0, n);
-        return n;
+        int nread;
+        if (target.hasArray()) {
+            char cbuf[] = target.array();
+            int off = target.arrayOffset() + target.position();
+            int len = target.remaining();
+            nread = this.read(cbuf, off, len);
+            if (nread > 0) {
+                target.position(target.position() + nread);
+            }
+        } else {
+            int remaining = target.remaining();
+            char cbuf[] = new char[Math.min(remaining, TRANSFER_BUFFER_SIZE)];
```

Comment:
That would be possible. It would help in cases where a large Reader is read into one or several relatively small off-heap CharBuffers, requiring multiple #read calls. This can only be done when the caller is able to work with only a partial input. I don't know how common this case is.

We could re-purpose #skipBuffer, it has the same maximum size (8192) but determined by a different constant (#maxSkipBufferSize instead of #TRANSFER_BUFFER_SIZE). That would likely require it to be renamed and maybe we should even remove #maxSkipBufferSize. We could also do the reallocation and growing similar as is currently done in #skip.

Reply:
Perhaps a static final `WORK_BUFFER_SIZE` could be added with value 8192 and `maxSkipBufferSize` and `TRANSFER_BUFFER_SIZE` replaced with that? Then `skipBuffer` could be renamed to `workBuffer` and used in both `read(CharBuffer)` and `skip(long)`. That shouldn't be a problem as both uses are in synchronized blocks. Also I suggest putting the declaration of `workBuffer` just below that of `lock` instead of lower down the file where `skipBuffer` is.

Lastly you mentioned C-style array declarations like `char buf[]`. As there are only four of these in the file it might be good to just go ahead and change them, I don't think that adds much noise or risk.

Reply:
Done. I left #transferTo(Writer) untouched for now. Firstly it is not already behind a synchronized. Secondly it writes so there is no need for repeated calls.